extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -0,0 +1,479 @@
+        final KafkaConsumer<byte[], byte[]> consumer = newConsumer()
+    ) {
+      appenderator = appenderator0;
+
+      final String topic = ioConfig.getStartPartitions().getTopic();
+
+      // Start up, set up initial offsets.
+      final Object metadata = appenderator.startJob();
+      final Map<Integer, Long> nextOffsets = Maps.newHashMap();
+      if (metadata == null) {
+        nextOffsets.putAll(ioConfig.getStartPartitions().getPartitionOffsetMap());
+      } else {
+        final Map<Integer, Long> metadataAsMap = (Map<Integer, Long>) metadata;
+        nextOffsets.putAll(metadataAsMap);
+
+        // Sanity check.
+        if (!nextOffsets.keySet().equals(ioConfig.getStartPartitions().getPartitionOffsetMap().keySet())) {
+          throw new ISE(
+              "WTF?! Restored partitions[%s] but expected partitions[%s]",
+              nextOffsets.keySet(),
+              ioConfig.getStartPartitions().getPartitionOffsetMap().keySet()
+          );
+        }
+      }
+
+      // Set up committer.
+      final Supplier<Committer> committerSupplier = new Supplier<Committer>()
+      {
+        @Override
+        public Committer get()
+        {
+          final Map<Integer, Long> snapshot = ImmutableMap.copyOf(nextOffsets);
+
+          return new Committer()
+          {
+            @Override
+            public Object getMetadata()
+            {
+              return snapshot;
+            }
+
+            @Override
+            public void run()
+            {
+              // Do nothing.
+            }
+          };
+        }
+      };
+
+      // Initialize consumer assignment.
+      final Set<Integer> assignment = Sets.newHashSet();
+      for (Map.Entry<Integer, Long> entry : nextOffsets.entrySet()) {
+        final long endOffset = ioConfig.getEndPartitions().getPartitionOffsetMap().get(entry.getKey());
+        if (entry.getValue() < endOffset) {
+          assignment.add(entry.getKey());
+        } else if (entry.getValue() == endOffset) {
+          log.info("Finished reading partition[%d].", entry.getKey());
+        } else {
+          throw new ISE(
+              "WTF?! Cannot start from offset[%,d] > endOffset[%,d]",
+              entry.getValue(),
+              endOffset
+          );
+        }
+      }
+
+      assignPartitions(consumer, topic, assignment);
+
+      // Seek to starting offsets.
+      for (final int partition : assignment) {
+        final long offset = nextOffsets.get(partition);
+        log.info("Seeking partition[%d] to offset[%,d].", partition, offset);
+        consumer.seek(new TopicPartition(topic, partition), offset);
+      }
+
+      // Main loop.
+      boolean stillReading = true;
+      while (stillReading) {
+        if (Thread.currentThread().isInterrupted()) {
+          throw new InterruptedException();
+        }
+
+        final ConsumerRecords<byte[], byte[]> records = consumer.poll(POLL_TIMEOUT);
```

Comment:
i guess poll(timeout) returns empty records on timeout, so the while loop is going to run indefinitely if there was no new data. we have to provide ways for early termination. i guess some of these things will need to be [re]visited as the "supervisor" takes shape.